RFC/src/RFC-0310_AssetImplementation.md

```diff
@@ -0,0 +1,607 @@
+and is only destroyed once the UTXO is spent.
+
+**New Consensus Rule**
+When receiving a transaction or block with an input with the `BURN` flag set, there must not exist a UTXO in the output set containing the same `unique_id`,`parent_public_key` pair, unless 
+that output has the `MINT` flag set.
+
+### Sidechain Checkpoints
+
+A sidechain checkpoint is a special type of NFT UTXO. Extra sidechain data can be specified in the `sidechain` field of the output features. Like 
+other NFT's, only one sidechain checkpoint UTXO with that unique_id can exist in the unspent set at one time. The conditions of 
+who can spend a checkpoint are determined by the TariScript on the checkpoint. In most cases it will be an `n of m` multisig condition 
+using the members of the Hotstuff committee, with `m` equal to the number of members of the committee and `n` equal to `m` minus 
+the number of failures the committee can tolerate.
+
+The unique_id can be a specificly chosen id, but can also be a random public key. 
+
+There is no limit to the number of different sidechain checkpoint tokens an asset has, but it will usually be one. 
+
+## New Base Node GRPC methods
+
+A base node should expose the following methods via GRPC:
+
+```protobuf
+
+  rpc GetTokens(GetTokensRequest) returns (stream GetTokensResponse);
+
+  message GetTokensRequest {
+    bytes asset_public_key = 1;
+    // Optionally get a set of specific unique_ids
+    repeated bytes unique_ids = 2;
+  }
+
+  message GetTokensResponse {
+    bytes unique_id = 1;
+    bytes asset_public_key = 2;
+    bytes owner_commitment = 3;
+    bytes mined_in_block = 4;
+    uint64 mined_height = 5;
+  }
+
+
+```
+
+## Validator Node Instructions
+
+> This section is specific to the Tari Validator Node. Other sidechains may make use of 
+> the base layer mechanisms for sidechains as well
+
+Callers may issue instructions to a validator node committee member using GRPC. Instructions take the form:
+
+```protobuf
+
+message Instruction{
+    // The public key of the asset being invoked
+    bytes asset_public_key =1;
+    // The template being invoked
+    uint32 template_id = 2;
+    // The method on the template being invoked.
+    string method =3;
+    // The position based arguments
+    repeated bytes args = 4;
+    // A commitment that exists on the base layer that identifies
+    // the caller. For example, the asset issuer could use the 
+    // commitment that holds the asset registration, while the owner
+    // could use the commitment on the UTXO containing the NFT
+    bytes caller_commitment = 4;
+    // A unique number to identify this instruction
+    uint64 id = 5;
+    // A comsig proving the caller owns the commitment in `caller_commitment`
+    // The challenge is the hash of the above fields, in order
+    bytes caller_sig = 6;
+    
+    // Fee fields TBD
+}
+
+```
+
+Upon receiving an instruction, a validator node must store it in its mempool and broadcast it to all other members of
```

Comment:
nit: Do we want to use the word mempool for the validator's memory storage of instructions?  The word mempool has become widely used in the blockchain's (aka baseLayer's) temp storage for transactions. Validator nodes technically don't deal with transactions but rather instructions. Perhals we should look into finding a unique name to help differentiate between the two?

Reply:
Mempool is used in this RFC already: https://rfc.tari.com/RFC-0340_VNConsensusOverview.html#implementation